esmvaltool/diag_scripts/ocean3d/utils.py

```diff
@@ -0,0 +1,245 @@
+"""APPLICATE/TRR Ocean Diagnostics."""
+import logging
+import os
+import shutil
+
+import cmocean.cm as cmo
+import matplotlib as mpl
+import matplotlib.cm as cm
+import numpy as np
+import palettable
+import pyproj
+import seawater as sw
+from cdo import Cdo
+
+mpl.use('agg')
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+class DiagnosticError(Exception):
+    """Error in diagnostic."""
+
+
+def genfilename(basedir,
+                variable=None,
+                mmodel=None,
+                region=None,
+                data_type=None,
+                extension=None,
+                basis='arctic_ocean'):
+    """Generate file name for the output data.
+
+    Parameters
+    ----------
+    basedir: str
+        base directory
+    variable: str
+        name of the variable
+    mmodel: str
+        name of the model
+    region: str
+        name of the region
+    data_type: str
+        type of the data, for example `timmean`
+    extention: str
+        fiel extention, for example `nc`
+    basis: str
+        basis name that can be used for series of
+        diagnostics
+
+    Returns
+    -------
+    ifilename: str
+        path to the file
+    """
+    nname = [basis, region, mmodel, variable, data_type]
+    nname_nonans = []
+    for i in nname:
+        if i:
+            nname_nonans.append(i)
+    basename = "_".join(nname_nonans)
+    if extension:
+        basename = basename + extension
+    ifilename = os.path.join(basedir, basename)
+    return ifilename
+
+
+def timmean(model_filenames, mmodel, cmor_var, diagworkdir,
+            observations='PHC'):
+    """Create time mean of input data with cdo.
```

Comment:
The preprocessor offers temporal averages that are lazy and parallelized. What is the advantage of this function?